src/stream/mod.rs

```diff
@@ -471,3 +476,206 @@ impl Stream {
+//         }
+//     }
+// }
+
+/// A wrapper around around [`Stream`], which implements [`AsyncRead`] and
+/// [`AsyncWrite`].
+///
+/// Both `poll_read` and `poll_write` calls allocate temporary buffers, which results in an
+/// additional overhead.
+struct PollStream<'a> {
+    stream: Arc<Stream>,
+
+    read_fut: Option<Pin<Box<dyn Future<Output = Result<usize>> + Send + 'a>>>,
+    write_fut: Option<Pin<Box<dyn Future<Output = Result<usize>> + Send + 'a>>>,
+    shutdown_fut: Option<Pin<Box<dyn Future<Output = Result<()>> + Send + 'a>>>,
+
+    read_buf: Vec<u8>,
+}
+
+impl PollStream<'_> {
+    /// Creates a new PollStream.
+    pub fn new(stream: Arc<Stream>) -> Self {
+        Self {
+            stream,
+            read_fut: None,
+            write_fut: None,
+            shutdown_fut: None,
+            read_buf: Vec::new(),
+        }
+    }
+    
+    /// Get back the inner stream.
+    pub fn into_inner(self) -> Arc<Stream> {
+        self.stream
+    }
+    
+    /// Obtain a clone of the inner stream.
+    pub fn clone_inner(&self) -> Arc<Stream> {
+        self.stream.clone()
+    }
+
+    /// stream_identifier returns the Stream identifier associated to the stream.
+    pub fn stream_identifier(&self) -> u16 {
+        self.stream.stream_identifier
+    }
+
+    /// buffered_amount returns the number of bytes of data currently queued to be sent over this stream.
+    pub fn buffered_amount(&self) -> usize {
+        self.stream.buffered_amount.load(Ordering::SeqCst)
+    }
+
+    /// buffered_amount_low_threshold returns the number of bytes of buffered outgoing data that is
+    /// considered "low." Defaults to 0.
+    pub fn buffered_amount_low_threshold(&self) -> usize {
+        self.stream.buffered_amount_low.load(Ordering::SeqCst)
+    }
+
+    /// get_num_bytes_in_reassembly_queue returns the number of bytes of data currently queued to
+    /// be read (once chunk is complete).
+    pub(crate) async fn get_num_bytes_in_reassembly_queue(&self) -> usize {
+        // No lock is required as it reads the size with atomic load function.
+        let reassembly_queue = self.stream.reassembly_queue.lock().await;
+        reassembly_queue.get_num_bytes()
+    }
+}
+
+impl AsyncRead for PollStream<'_> {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut ReadBuf<'_>,
+    ) -> Poll<io::Result<()>> {
+        if buf.remaining() == 0 {
+            return Poll::Ready(Ok(()));
+        }
+
+        let fut = match self.read_fut.as_mut() {
+            Some(fut) => fut,
+            None => {
+
+                // read into a temporary buffer because `buf` has an unonymous lifetime, which can be
+                // shorter than the lifetime of `read_fut`.
+                let stream = self.stream.clone();
+                let temp_buf = Vec::with_capacity(buf.capacity());
+                self.read_fut.get_or_insert(Box::pin(( move || {
```

Comment:
@tomaka this line produces the following error:
![image](https://user-images.githubusercontent.com/1282182/163572478-32d04748-ae3e-47e5-96d3-9ac88a9e4429.png)

in short, Rust can't figure out lifetime for `read` Future. 

NOTE: if you remove the clojure, the problem becomes:`stream` and `temp_buf`, whos lifetime becomes too short "while they are still in use" by Future. Hence I've added clojure to prolong their lifetime.

Reply:
I think the issue is simply that you're calling `stream.read` and returning the `Future` that `stream.read` returns.

Something like this should work:

```rust
Box::pin(async move {   // Note the async
    stream.read(temp_buf.as_mut_slice());
    temp_buf
})
```

Reply:
Probably more like:

```rust
let n = stream.read(temp_buf.as_mut_slice()).await;
temp_buf.truncate(n);
temp_buf
```

But you get the idea

Reply:
awesome! thanks